### ISSUE TITLE
Notify users about password expiration (based on PSO)

### ISSUE THREAD

**Issue by schmidtsva** [LDAPAccountManager/lam#387]
Hello,

We would like to create a cron job that notifies the users in our Samba4 AD when their password expires. We use Password Settings Objects (PSO) to set the maximum age of passwords.

As far as I could determine, LAM only calculates the password expiration based on the global policy and not based on the PSOs. Would it be possible to take the PSOs into account when calculating the expiration date?

Following example: The global domain policy is 180 days. However, the user “user.dd02” has a maximum password age of 10 days via PSO. The user's password was changed on 2024-11-08.

```
root@pve-addc1-111:~# samba-tool domain passwordsettings pso show-user user.dd02
The following PSO settings apply to user 'user.dd02'.

Password information for PSO 'pso_max-pwd-age_10day'

Precedence (lowest is best): 5
Password complexity: on
Store plaintext passwords: off
Password history length: 24
Minimum password length: 7
Minimum password age (days): 0
Maximum password age (days): 10    <<<-----<<<-----<<<----- max. 10 days
Account lockout duration (mins): 30
Account lockout threshold (attempts): 3
Reset account lockout after (mins): 30

PSO applies to user via group membership.
root@pve-addc1-111:~#
```

However, LAM calculates an expiration date after 180 days on 2025-05-07.

```
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Checking CN=user.dd02,OU=Users,OU=Dresden,DC=pve-ad,DC=lan
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Last password change on 2024-11-08
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Number of days before warning 10
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: PwdLastSet: 133755373580327820
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Expiration raw value: 133910893580327820
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Password expires on 2025-05-07
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: Password notification on 2025-04-27 13:02
Nov 08 13:50:02 pve-lam-112 php[94104]: LDAP Account Manager ( - - CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan) - DEBUG: CN=user.dd02,OU=Users,OU=Dresden,DC=pve-ad,DC=lan does not need notification yet.
```

Best Regards

Christoph Schmidt

**Comment by gruberroland** [comment]
Hi Christoph,

thanks for the info, is there any link from the user to the PSO that can be seen in LDAP? Do you see something in the LDAP tree view?

**Comment by schmidtsva** [comment]
Hi Roland,

yes, as far as I have found out, there is the "constructed" attribute ``MsDS-UserPasswordExpiryTimeComputed``, which can be queried for this purpose. The ``msDS-ResultantPSO`` attribute shows the PSO to which the user is linked. I just tried that:

```
root@pve-addc1-111:~# ldapsearch -x -ZZ -H ldap://pve-addc1-111 -b "DC=pve-ad,DC=lan" -D "CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan" -W "(&(objectCategory=Person)(objectClass=User))" sn givenname displayName sAMAccountName msDS-UserPasswordExpiryTimeComputed msDS-ResultantPSO mail
Enter LDAP Password: 
# extended LDIF
#
# LDAPv3
# base <DC=pve-ad,DC=lan> with scope subtree
# filter: (&(objectCategory=Person)(objectClass=User))
# requesting: sn givenname displayName sAMAccountName msDS-UserPasswordExpiryTimeComputed msDS-ResultantPSO mail 
#

# LDAP Bind, Users, pve-ad.lan
dn: CN=LDAP Bind,CN=Users,DC=pve-ad,DC=lan
sn: Bind
givenName: LDAP
displayName: LDAP Bind
sAMAccountName: ldapbind
msDS-UserPasswordExpiryTimeComputed: 9223372036854775807

# Administrator, Users, pve-ad.lan
dn: CN=Administrator,CN=Users,DC=pve-ad,DC=lan
sAMAccountName: Administrator
msDS-UserPasswordExpiryTimeComputed: 133871153296412020

# DD03 User, Users, Dresden, pve-ad.lan
dn: CN=DD03 User,OU=Users,OU=Dresden,DC=pve-ad,DC=lan
sn: User
givenName: DD03
displayName: DD03 User
sAMAccountName: user.dd03
mail: user.dd03@pve-ad.lan
msDS-UserPasswordExpiryTimeComputed: 133512626596007910
msDS-ResultantPSO: CN=pso_max-pwd-age_10day,CN=Password Settings Container,CN=System,DC=pve-ad,DC=lan

# user.dd02, Users, Dresden, pve-ad.lan
dn: CN=user.dd02,OU=Users,OU=Dresden,DC=pve-ad,DC=lan
sAMAccountName: user.dd02
mail: user.dd02@pve-ad.lan
msDS-UserPasswordExpiryTimeComputed: 133764013580327820
msDS-ResultantPSO: CN=pso_max-pwd-age_10day,CN=Password Settings Container,CN=System,DC=pve-ad,DC=lan

# user.dd01, Users, Dresden, pve-ad.lan
dn: CN=user.dd01,OU=Users,OU=Dresden,DC=pve-ad,DC=lan
sAMAccountName: user.dd01
displayName: user.dd01
mail: user.dd01@pve-ad.lan
msDS-UserPasswordExpiryTimeComputed: 133720852535380260
msDS-ResultantPSO: CN=pso_max-pwd-age_1day,CN=Password Settings Container,CN=System,DC=pve-ad,DC=lan
...
```

When I convert the date/time value of the user user.dd02, the correct expiration time of the password is displayed.

```
C:\Users\administrator>w32tm /ntte 133764013580327820
154819 11:02:38.0327820 - 18.11.2024 12:02:38

C:\Users\administrator>
```

**Comment by gruberroland** [comment]
Thanks, we will update the job to use MsDS-UserPasswordExpiryTimeComputed then. Thanks a lot for all the details.

**Comment by gruberroland** [comment]
Will be published with 9.0
